src/economy/shipping_schedule.cc

```diff
@@ -0,0 +1,1715 @@
+
+	for (PortDock* pd : fleet_.get_ports()) {
+		for (auto it = pd->waiting_.begin(); it != pd->waiting_.end();) {
+			if (it->destination_dock_.serial() == dock->serial()) {
+				sslog("found a shippingitem in port %u\n", pd->serial());
+				it->set_location(game, pd->warehouse_);
+				it->end_shipping(game);
+				it = pd->waiting_.erase(it);
+			} else {
+				++it;
+			}
+		}
+	}
+
+	for (Ship* ship : ships_heading_there) {
+		for (ShippingItem& si : ship->items_) {
+			if (si.destination_dock_.serial() == dock->serial()) {
+				sslog("found a shippingitem on %s\n", ship->get_shipname().c_str());
+				si.destination_dock_ = ship->get_destination();
+			}
+		}
+	}
+	sslog("--- port_removed maintenance complete ---\n\n");
+}
+
+void ShippingSchedule::ship_removed(const Game&, Ship* ship) {
+	auto it = plans_.find(ship);
+	assert(it != plans_.end());
+	plans_.erase(it);
+
+	auto i = last_actual_duration_recalculation_.find(ship);
+	if (i != last_actual_duration_recalculation_.end()) {
+		last_actual_duration_recalculation_.erase(i);
+	}
+
+	// Handling any items that were intended to be transported by this ship
+	// is deferred to the next call to update()
+}
+
+void ShippingSchedule::ship_added(Game& game, Ship& s) {
+	sslog("\nShippingSchedule::ship_added (%s)\n", s.get_shipname().c_str());
+	assert(!s.get_destination());
+	plans_[&s] = ShipPlan();
+	last_actual_duration_recalculation_[&s] = game.get_gametime();
+	if (fleet_.get_ports().empty()) {
+#ifndef NDEBUG
+		for (ShippingItem& si : s.items_) {
+			assert(!si.destination_dock_.is_set());
+		}
+#endif
+		sslog("No ports!\n\n");
+		return;
+	}
+
+	PortDock* closest = nullptr;
+	int32_t dist = 0;
+	for (PortDock* dock : fleet_.get_ports()) {
+		Path path;
+		int32_t d = -1;
+		s.calculate_sea_route(game, *dock, &path);
+		game.map().calc_cost(path, &d, nullptr);
+		assert(d >= 0);
+		if (!closest || d < dist) {
+			dist = d;
+			closest = dock;
+		}
+	}
+	assert(closest);
+	for (ShippingItem& si : s.items_) {
+		si.destination_dock_ = closest;
+	}
+	plans_[&s].push_back(SchedulingState(closest, false, dist));
+	s.set_destination(game, closest);
+	sslog("Sent to %u\n\n", closest->serial());
+}
+
+void ShippingSchedule::port_added(Game& game, PortDock& dock) {
+	sslog("\nShippingSchedule::port_added (%u)\n", dock.serial());
+	if (fleet_.count_ports() > 1) {
+		// nothing to do currently
+		sslog("nothing to do\n\n");
+		return;
+	}
+	// All ships are most likely panicking because they have
+	// no destination. Send them all to the new port.
+	for (Ship* ship : fleet_.get_ships()) {
+		assert(!ship->get_destination());
+		sslog("Rerouting %s there\n", ship->get_shipname().c_str());
+		ship->set_destination(game, &dock);
+		for (ShippingItem& si : ship->items_) {
+			assert(!si.destination_dock_.is_set());
+			si.destination_dock_ = &dock;
+		}
+	}
+	sslog("--- port_added maintenance complete ---\n\n");
+}
+
+Duration ShippingSchedule::update(Game& game) {
```

Comment:
Yelp, this is a monster of a function with > 1000 lines. Outsch.

Reply:
I would very much prefer to split off 6 functions for the 6 passes (or even more, e.g. the fifth pass qualifies for 4 functions) but unfortunately the passes need to communicate with each other. So the subfunctions would need lots of return values or mutable pointers which would be a questionable gain…

Reply:
Once I have a _lot_ of time Ill try that ... :-)